src/solidity-frontend/pattern_check.cpp

```diff
@@ -0,0 +1,119 @@
+#include <solidity-frontend/pattern_check.h>
+
+pattern_checker::pattern_checker(const nlohmann::json &_ast_nodes,
+    const std::string &_target_func, const messaget &msg):
+    ast_nodes(_ast_nodes),
+    target_func(_target_func),
+    msg(msg)
+{
+}
+
+bool pattern_checker::pattern_checker::do_pattern_check()
+{
+  // add more functions here to perform more pattern-based checks
+  //print_json_element(ast_nodes);
+  printf("Checking function %s ...\n", target_func.c_str());
+
+  unsigned index = 0;
+  nlohmann::json::const_iterator itr = ast_nodes.begin();
+  for (; itr != ast_nodes.end(); ++itr, ++index)
+  {
+    if ((*itr).contains("kind") && (*itr).contains("nodeType") && (*itr).contains("name"))
+    {
+      // locate the target function
+      if ((*itr)["kind"].get<std::string>() == "function" &&
+          (*itr)["nodeType"].get<std::string>() == "FunctionDefinition" &&
+          (*itr)["name"].get<std::string>() == target_func)
+      {
+        return start_pattern_based_check(*itr);
+      }
+    }
+  }
+
+  return false;
+}
+
+bool pattern_checker::start_pattern_based_check(const nlohmann::json &func)
+{
+  // SWC-115: Authorization through tx.origin
+  check_authorization_through_tx_origin(func);
+  return false;
+}
+
+void pattern_checker::check_authorization_through_tx_origin(const nlohmann::json &func)
+{
+  // looking for the pattern require(tx.origin == <VarDeclReference>)
+  const nlohmann::json &body_stmt = func["body"]["statements"];
+  msg.progress("  - Pattern-based checking: SWC-115 Authorization through tx.origin");
+  printf("statements in function body array ... \n");
+  //print_json_element(body_stmt);
+
+  unsigned index = 0;
+  nlohmann::json::const_iterator itr = body_stmt.begin();
+  for (; itr != body_stmt.end(); ++itr, ++index)
+  {
+    printf("@@ checking body stmt %u\n", index);
+    if (itr->contains("nodeType"))
+    {
+      if ((*itr)["nodeType"].get<std::string>() == "ExpressionStatement")
+      {
+        const nlohmann::json &expr = (*itr)["expression"];
+        //print_json_element(expr);
+        if (expr["nodeType"] == "FunctionCall")
+        {
+          if (expr["kind"] == "functionCall")
+          {
+            // require(.) authorization
+            if (expr["expression"]["nodeType"].get<std::string>() == "Identifier")
```

Comment:
Try to break this large if-statement into smaller methods.

Reply:
Done. 
Split into the following functions:
* check_authorization_through_tx_origin
* check_require_call
* check_require_argument
* check_tx_origin